NetKet/Hamiltonian/graph_hamiltonian.hpp

```diff
@@ -0,0 +1,120 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_BOND_HAMILTONIAN_CC
+#define NETKET_BOND_HAMILTONIAN_CC
+
+#include <Eigen/Dense>
+#include <iostream>  // TODO remove
+#include <vector>
+#include "Utils/json_helper.hpp"
+#include "local_operator.hpp"
+
+namespace netket {
+
+// BondHamiltonian on an arbitrary graph
+template <class G>
+class GraphHamiltonian : public AbstractHamiltonian {
+  std::vector<LocalOperator> operators_;
+  Hilbert hilbert_;
+
+  // Arbitrary graph
+  const G &graph_;
```

Comment:
Why a reference? I would suggest you use a pointer instead as this would make `GraphHamiltonian` copy and move assignable (which is always a nice thing to have).

Reply:
I used a reference to be consistent with how graphs are stored in the BosonHubbard, Heisenberg, and Ising hamiltonians. If it's something that we want, should I also change it for those Hamiltonians?

Reply:
Reference is fine